server/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java

```diff
@@ -589,6 +610,112 @@ void recoverFilesFromSourceAndSnapshot(ShardRecoveryPlan shardRecoveryPlan,
+            listener.onResponse(Collections.emptyList());
+            return;
+        }
+
+        new SnapshotRecoverFileRequestsSender(shardRecoveryPlan, listener).start();
+    }
+
+    class SnapshotRecoverFileRequestsSender {
+        private final ShardRecoveryPlan.SnapshotFilesToRecover snapshotFilesToRecover;
+        final ActionListener<List<StoreFileMetadata>> listener;
+        final CountDown countDown;
+        final BlockingQueue<BlobStoreIndexShardSnapshot.FileInfo> pendingSnapshotFilesToRecover;
+        final AtomicBoolean cancelled = new AtomicBoolean();
+        List<StoreFileMetadata> filesFailedToDownloadFromSnapshot;
+
+        SnapshotRecoverFileRequestsSender(ShardRecoveryPlan shardRecoveryPlan, ActionListener<List<StoreFileMetadata>> listener) {
+            this.snapshotFilesToRecover = shardRecoveryPlan.getSnapshotFilesToRecover();
+            this.listener = listener;
+            this.countDown = new CountDown(shardRecoveryPlan.getSnapshotFilesToRecover().size());
+            this.pendingSnapshotFilesToRecover =
+                new LinkedBlockingQueue<>(shardRecoveryPlan.getSnapshotFilesToRecover().getSnapshotFiles());
+        }
+
+        void start() {
+            for (int i = 0; i < maxConcurrentSnapshotFileDownloads; i++) {
+                sendRequest();
+            }
+        }
+
+        void sendRequest() {
+            BlobStoreIndexShardSnapshot.FileInfo snapshotFileToRecover = pendingSnapshotFilesToRecover.poll();
+            if (snapshotFileToRecover == null) {
+                return;
+            }
+
+            try {
+                cancellableThreads.checkForCancel();
+
+                ActionListener<Void> sendRequestListener = new ActionListener<>() {
+                    @Override
+                    public void onResponse(Void unused) {
+                        onRequestCompletion(snapshotFileToRecover.metadata(), null);
+                    }
+
+                    @Override
+                    public void onFailure(Exception e) {
+                        logger.warn(new ParameterizedMessage("failed to recover file [{}] from snapshot, " +
+                            "will recover from primary instead", snapshotFileToRecover.metadata()), e);
+                        onRequestCompletion(snapshotFileToRecover.metadata(), e);
+                    }
+                };
+
+                recoveryTarget.downloadSnapshotFile(
+                    snapshotFilesToRecover.getRepository(),
+                    snapshotFilesToRecover.getIndexId(),
+                    snapshotFileToRecover,
+                    ActionListener.runAfter(sendRequestListener, this::sendRequest)
+                );
+            } catch (Exception e) {
+                onFailure(e);
+            }
+        }
+
+        void onFailure(Exception e) {
+            if (cancelled.compareAndSet(false, true)) {
+                pendingSnapshotFilesToRecover.clear();
+                listener.onFailure(e);
```

Comment:
If an exception is thrown during `sendRequest` but not on the first file, I think we risk still being in the process of downloading files concurrently with the subsequent "normal" phase1 of recovery? I am not 

Also, I would think we would need to add all the files we did not download to `filesFailedToDownloadFromSnapshot` in order to make the subsequent normal file based recovery copy them over?

Reply:
Good catch, I think I should be more explicit here and cancel only after an `ExecutionCancelledException` and handle the other case as a regular failure, wdyt?

Reply:
I think it would be preferable to complete the listener only after all the outstanding requests have finished. Neither option is great, the download might delay the cancellation for a long time, but I think we'll encounter a nasty surprise at some point in the future if we have downloads still happening while we're sending file chunks.

Reply:
Do you mean if the recovery is retried? In that case the tmp file names should be different. But I agree, let's be safe here.

Reply:
Actually maybe there is an issue with temp-file reuse anyway: if we complete the download of a file but then fail to acknowledge it to the source then we would try and copy over the same file using the same temp file name, which would throw an `IllegalStateException` in `MultiFileWriter`.

Maybe we just retry until we get that acknowledgement today (e.g. we retry through `NodeDisconnectedException` and `CircuitBreakingException` and other likely candidates) but are we sure that covers everything?

Reply:
> Do you mean if the recovery is retried?

Oh right yes the only exception we care about here is one indicating that the recovery is cancelled so we won't do any more work.

Reply:
Addressed in https://github.com/elastic/elasticsearch/pull/76237/commits/09fa0add0295a432fea6816ceddcd58070036e9a